### DISCUSSION TITLE
Adding extra images to an item

### DISCUSSION THREAD

**Discussion by azuravian** [benjaminjonard/koillection#1150] [Q&A] [answered]
I'm using the API to create items and add data to them.  I'm having no problem using the /api/items/{id}/image endpoint to upload the primary image for the item, but I also want to upload additional images as data of the item.  I've tried so many things I can't even remember them all.

From the API page, it appears that to load a data image, I need to:
1. use the /api/data/{id}/image endpoint with POST.  This endpoint requires you to provide a datum id, so I'm assuming I need to
2. use /api/data endpoint with POST to create the datum returning the id.  I then try to use the data image endpoint from step 1
3. using /api/data/{id from step 2}/image with POST.
4. This returns a 201, showing that it is reporting the correct datum id, but there are no image paths in the response.  
 
I end up with an Image added to the item that has the correct label, but just the default template image, as opposed to the one I submitted in step 2.  When I check the uploads folder, no new images were created.

Any ideas what I'm doing wrong?

**Comment by benjaminjonard** [comment, marked as answer]
Hello, the process is as you described. Maybe there is something missing in your data. Here's a working example: 

1/ Create the data with `POST https://localhost:81/api/data`

```
{
  "item": "/api/items/018ffc08-c6ff-78ca-95aa-62ccd206808d",  
  "type": "image",
  "label": "Test image"
}
```

it should returns something like this 
```
{
	"id": "018ffc0e-1a1d-7cb2-a0c3-c6b6c5d4ac35",
	"item": "\/api\/items\/018ffc08-c6ff-78ca-95aa-62ccd206808d",
	"collection": null,
	"type": "image",
	"label": "Test image",
	"value": null,
	"position": null,
	"image": null,
	"imageSmallThumbnail": null,
	"imageLargeThumbnail": null,
	"file": null,
	"video": null,
	"originalFilename": null,
	"choiceList": null,
	"owner": "\/api\/users\/96d5e71f-e034-4675-a6bf-77affd459f58",
	"visibility": "public",
	"parentVisibility": "public",
	"finalVisibility": "public",
	"createdAt": "2024-06-09T10:12:13+02:00",
	"updatedAt": null
}
```

2/ Upload the image with `POST https://localhost:81/api/data/018ffc0e-1a1d-7cb2-a0c3-c6b6c5d4ac35/image`

You need to use multipart/form-data as Content-Type.

Here is a screenshot from Insomnia (the tool I use to test the API)
![Screenshot from 2024-06-09 11-13-04](https://github.com/benjaminjonard/koillection/assets/20560781/494684a7-27fc-4b82-927f-d1d52b3e8b9c)


If it's not helping, could you share the code you use for the upload ?

**Comment by azuravian** [comment]
Ah, your last image showed me what I was doing wrong.  I was still using "file" instead of "fileImage".

Could you possibly show me the correct way to use the API to add a list to an item.  When I try to send an array, it tells me it must be a string, when I format the array like a string (which is what I get when using GET on a list), it gives me a 201, but the entry is null.  When I then try to navigate to the item in the web interface, I get the following:

![image](https://github.com/benjaminjonard/koillection/assets/7379812/0f5e3543-79f3-4048-8e28-33f7f555371d)

I can edit the item and see that the list was added and labeled correctly, but there are no list elements present.  If I add at least one manually, the error in the image above goes away.

**Comment by benjaminjonard** [comment]
For lists the value must be a JSON valid string like  `["Value 1","Value 2","Value 3"]`.  The problem may be that double quotes are mandatory in json

**Comment by azuravian** [comment]
You got it.  I just figured that out and came to post that, but you were a couple minutes ahead of me.  Everything working now.  Thank you so much.
